src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,763 @@
+ * questions.
+ *
+ */
+
+#include "jfr/support/jfrThreadLocal.hpp"
+#include "memory/resourceArea.hpp"
+#include "runtime/atomic.hpp"
+#include "runtime/orderAccess.hpp"
+#include "utilities/ticks.hpp"
+#include "jfr/periodic/sampling/jfrCPUTimeThreadSampler.hpp"
+
+#if defined(LINUX)
+
+#include "jfr/recorder/jfrRecorder.hpp"
+#include "jfr/recorder/service/jfrRecorderService.hpp"
+#include "jfr/utilities/jfrThreadIterator.hpp"
+#include "jfr/utilities/jfrTypes.hpp"
+#include "jfr/utilities/jfrTime.hpp"
+#include "jfrfiles/jfrEventClasses.hpp"
+#include "runtime/javaThread.hpp"
+#include "runtime/osThread.hpp"
+#include "runtime/safepointMechanism.hpp"
+#include "runtime/safepointMechanism.inline.hpp"
+#include "runtime/threadSMR.hpp"
+#include "runtime/vmOperation.hpp"
+#include "runtime/vmThread.hpp"
+
+#include "signals_posix.hpp"
+
+static const int64_t AUTOADAPT_INTERVAL_MS = 100;
+
+static bool is_excluded(JavaThread* thread) {
+  return thread->is_hidden_from_external_view() || thread->jfr_thread_local()->is_excluded();
+}
+
+static JavaThread* get_java_thread_if_valid() {
+  Thread* raw_thread = Thread::current_or_null_safe();
+  JavaThread* jt;
+  if (raw_thread == nullptr || !raw_thread->is_Java_thread()) { // this can happen due to the high level of parralelism
+    return nullptr;
+  }
+  if ((jt = JavaThread::cast(raw_thread))->is_exiting()) {
+    return nullptr;
+  }
+
+  if (is_excluded(jt)) {
+    return nullptr;
+  }
+  return jt;
+}
+
+JfrCPUTimeTraceQueue::JfrCPUTimeTraceQueue(u4 capacity) :
+  _capacity(capacity), _head(0), _lost_samples(0) {
+  _data = JfrCHeapObj::new_array<JfrCPUTimeSampleRequest>(capacity);
+}
+
+JfrCPUTimeTraceQueue::~JfrCPUTimeTraceQueue() {
+  JfrCHeapObj::free(_data, _capacity * sizeof(JfrCPUTimeSampleRequest));
+}
+
+bool JfrCPUTimeTraceQueue::enqueue(JfrCPUTimeSampleRequest& request) {
+  assert(JavaThread::current()->jfr_thread_local()->is_cpu_time_jfr_enqueue_locked(), "invariant");
+  u4 elementIndex;
+  do {
+    elementIndex = Atomic::load_acquire(&_head);
+    if (elementIndex >= _capacity) {
+      return false;
+    }
+  } while (Atomic::cmpxchg(&_head, elementIndex, elementIndex + 1) != elementIndex);
+  _data[elementIndex] = request;
+  return true;
+}
+
+JfrCPUTimeSampleRequest& JfrCPUTimeTraceQueue::at(u4 index) {
+  assert(index < _head, "invariant");
+  return _data[index];
+}
+
+volatile u4 _lost_samples_sum = 0;
+
+u4 JfrCPUTimeTraceQueue::size() const {
+  return Atomic::load(&_head);
+}
+
+void JfrCPUTimeTraceQueue::set_size(u4 size) {
+  Atomic::release_store(&_head, size);
+}
+
+u4 JfrCPUTimeTraceQueue::capacity() const {
+  return _capacity;
+}
+
+void JfrCPUTimeTraceQueue::set_capacity(u4 capacity) {
+  _head = 0;
+  JfrCPUTimeSampleRequest* new_data = JfrCHeapObj::new_array<JfrCPUTimeSampleRequest>(capacity);
+  JfrCHeapObj::free(_data, _capacity * sizeof(JfrCPUTimeSampleRequest));
+  _data = new_data;
+  _capacity = capacity;
+}
```

Comment:
I assume there is a lock protecting this so it happens atomically?

Reply:
This happens before the signal handler is attached to thread. So it does happen before any parallelism is introduced on thread creation.

Reply:
I'm missing the big picture here unfortunately. This looks like it can get called repeatedly as needed to change capacity. Are you saying it only gets called once before we create the sampler thread? Is the concurrency model described somewhere?